### PR TITLE
Deprecate k_mem_pool API, remove sys_mem_pool allocator

### LINKED ISSUE #24358: deprecate and remove old k_mem_pool / sys_mem_pool APIs
We have replaced the mem pool APIs with the new k_heap/sys_heap APIs. We don't need to keep maintaining two APIs for doing the same thing, and should aim to remove the k_mem_pool and sys_mem_pool APIs from the kernel.

The first step is to `__deprecate` them all, and change any use within Zephyr to use k_heap/sys_heap instead.

The `k_pipe_block_put()` API will need to be adjusted as it takes a `k_mem_block` as an argument. I couldn't find any other public APIs which would need to be changed.

### PR DESCRIPTION
**Pull request by andyross**
The k_heap/sys_heap code has been default for a release, so it's time to deprecate the old allocator.

This eliminates the older CONFIG_MEM_POOL_HEAP_BACKEND kconfig, effectively making it true always and removing the ability for applications to request the older pool engine.  It then deprecates the older APIs, leaving them in place for internal users as a "Z" API.

There are still a few users of the older API internally that will need cleanup (**edit: not anymore, it's 100% gone in the current version**).  I'll submit those separately I think, I'm not sure if those are in scope for 2.4 or not:

* k_malloc and the system pool are still implemented in terms of mem_pool, even though those are just wrappers now.  It should be directly on top of a k_heap.

* The per-thread resource pools likewise are mem_pools still.  This should be k_heap's instead.

* Mailbox has a API that will retrieve a message into a newly-allocated heap block.  This API is weird and itself deprecated, but if we decide we actually want to preserve the functionality it will need some kind of new variant.

* Similarly pipe has block usage internally.

And there are two places where the underlying sys_heap allocator still gets used:

* The minimal libc malloc() implementation is on top of sys_heap instead of k_malloc.  I'm not sure why, but this needs some kind of port.

* The lib/gui/lvgl code has an internal sys_mem_pool and needs a port to sys_heap (not hard to write, but I have no idea how to test it)

Fixes #24358

### CODE REVIEW

**Comment by andyross** [comment]
Ugh, it collides.  Sorry, I did this on a week-stale master.  Will get that cleaned up.

**Review by nashif — DISMISSED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by pabigot** [comment]
I'm not clear on how #24358 is a bug, let alone having a release-blocking priority, except as a means of getting this late change into 2.4.0.  And I've been told before that stage of release cycle is not a valid reason for rejecting a PR.

But from policy discussion heard at the last TSC regarding criteria for accepting late addition of new boards I need to ask: If a PR like this came three days before release from anybody other than a platinum member company, would people be seriously considering merging it into the release when it passes?

**Comment by andrewboie** [comment]
> I'm not clear on how #24358 is a bug

It's a bug because we have two mechanisms in the kernel which do the same thing, and we do not want to maintain both, particularly not in the second LTS release.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by nashif** [comment]
> If a PR like this came three days before release from anybody other than a platinum member company, would people be seriously considering merging it into the release when it passes?

seriously? What exactly is the purpose of creating releases and having almost 4 weeks of bug fix and stabilization period if we do not address issues like this? 

This was a bug when it was created in April, move to an Enhancement because it did not happen for 2.3 with clear documentation in the issue itself that it should go into 2.4. This had nothing to do with membership level and it is disappointing to hear that TBH. This is not something someone is pushing for to get a feature in late in the game, this actually removes duplication and confusion. So there is no direct and "personal" gain from doing this.

**Review by pabigot — COMMENTED** [review]
Looking closely it seems the impact would only be on applications that explicitly selected `CONFIG_MEM_POOL_HEAP_BACKEND=n`, so once the missed in-tree uses are fixed and it passes I don't see any technical reason it shouldn't go in.

**Comment by pabigot** [comment]
Seriously, the need to do this has been known for months.  Why is it being done so late, and why is it being forced/enabled by giving #24358 a release-blocking priority?  Zephyr's ability to plan what goes into a release, and track its progress, is disappointingly weak.

I do feel that part of the criteria for accepting PRs in the stabilization period should be the risk of destabilizing, invalidating all the testing that's been done, and introducing build warnings in out of tree applications and downstream forks that have been faithfully tracking the release and suddenly have to make changes to accommodate a last minute API deprecation.

That there are so many remaining in-tree uses that fail CI because of the new deprecation suggests this isn't a baseless concern.

And I remain very skeptical that this would be getting this level of attention and support if a random community contributor had proposed it at this point.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
Just noticed: this needs to deprecate all the `sys_mem_pool` APIs as well as they are superseded by `sys_heap`.

If you want to remove the `sys_mem_pool` stuff in another PR please remove the "Fixes" line in the PR description. I'd rather it was removed here though.

This PR was discussed in the TSC today, the upshot is that we will wait until the merge window opens for 2.5 before merging. And then completely remove for 2.6, with a documented exception to the two release rule. The 2.4 release notes will indicate these APIs are now considered deprecated even though they will not be tagged with `__deprecated` until 2.5.

**Comment by andrewboie** [comment]
This also needs to deprecate `k_pipe_block_put()`

AFAICT the full set of APIs to deprecate is:
  * `k_mbox_data_block_get()`
  * `k_pipe_block_put()`
  * `K_MEM_POOL_DEFINE()`
  * `k_mem_pool_alloc()`
  * `k_mem_pool_free()`
  * `k_mem_pool_free_id()`
  * `SYS_MEM_POOL_DEFINE()`
  * `sys_mem_pool_init()`
  * `sys_mem_pool_alloc()`
  * `sys_mem_pool_free()`
  * `sys_mem_pool_try_expand_inplace()`

Since this is now in scope for early 2.5 and not 2.4, I would prefer if these deprecation PRs converted internal kernel usage to k_heap / sys_heap rather than introducing these z_ variants.

**Comment by carlescufi** [comment]
@andyross @andrewboie could we pick this up now that 2.4 has been released? let's not wait until the next release is close yet again.

**Comment by MaureenHelm** [comment]
> @andyross @andrewboie could we pick this up now that 2.4 has been released? let's not wait until the next release is close yet again.

Let's get this moving again

**Comment by nashif** [comment]
> Let's get this moving again

@andyross  already working on this.

**Comment by andyross** [comment]
Resurrect this guy with a more invasive change.  Contains the old deprecation, and also a complete removal of the older sys_mem_pool allocator backend.  That code (and a bunch of tests specific to it) is now gone, the only allocator we ship is the new one.

The compatibility API (which implements the old scheme in terms of the new allocator) remains in place as "z_mem_pool" for the benefit of a few in-tree spots that still use it and are difficult for me to test directly (mostly driver code for devices I don't have).  We can address these piecewise as we find maintainers able to validate, that stuff is all boilerplate.

And this does not contain a patch that actually removes the deprecated APIs.  I'm still not clear on the project management intent there as to whether that should happen this release or next.

**Comment by nashif** [comment]
```

/workdir/zephyr/tests/kernel/mem_protect/syscalls/src/main.c:392:2: error: 'z_thread_resource_pool_assign' is deprecated [-Werror=deprecated-declarations]
--



```

**Comment by andyross** [comment]
> did you miss one?

Yeah, last minute goof.  That function is an internal wrapper, it's not supposed to be deprecated.  This happened because there had been a "temporarily undeprecated" k_* function there at one point (i.e. to suppress the deprecation warnings while I worked on somethign else), which worked by commenting the function out with a C++ comment I knew would get caught later (e.g. "//__deprecated // FIXME reenable"), which did indeed get caught in a final checkpatch.  But then I blindly reenabled it forgetting that this was now an internal function that didn't need to be deprecated and pushed before re-running the whole sanitycheck.

Probably too much detail.  Easy fix.  Will push as soon as my rebased tree completes its check.

**Comment by andyross** [comment]
I'd been assuming those weird BSIM errors (in code that has nothing to do with mem_pool) were an upstream thing, but they turned out to be a latent bug with a header (missing stdint.h) that the removal of mempool_sys.h exposed.  Fixed.

File: include/kernel.h
Comment (andrewboie) [inline include/kernel.h:619]: this will break existing users.
better I think to leave this function name as `k_thread_resource_pool_assign` and mark as `__deprecated`

**Review by tyrantlizard77 — APPROVED** [review]
one nit about an API name change. rest looks fine.

I think we will need to file bugs or otherwise figure out what to do about the remaining uses of z_mem_pool_* inside the kernel so that we can remove all the z_mem_pool code entirely for LTS2, but we have time.

File: drivers/usb/device/usb_dc_kinetis.c
Comment (nashif) [inline drivers/usb/device/usb_dc_kinetis.c:83]: shouldn we be using K_HEAP_DEFINE here?

File: drivers/usb/device/usb_dc_kinetis.c
Comment (nashif) [inline drivers/usb/device/usb_dc_kinetis.c:356]: why are we using this internal API and not k_heap_free?

**Review by nashif — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_kinetis.c
Comment (andyross) [reply drivers/usb/device/usb_dc_kinetis.c:83]: Yes, but I don't feel comfortable doing the port without a device with this USB controller to validate.  The wrappers are near-zero-size and very low risk.  Changing API usage blindly seems like more downside than up.

Comment (jfischer-phytec-iot) [reply drivers/usb/device/usb_dc_kinetis.c:83]: I did https://github.com/jfischer-phytec-iot/zephyr/commit/7f482950a000072d2b3ba97b080d9b8de0fed358 just for fun. IMHO K_HEAP_DEFINE is not an easy replacement for K_MEM_POOL_DEFINE as the overhead for the heap management has to be considered. However we want to remove this intermediate buffers anyway, so I am fine to do minimal changes and use Z_MEM_POOL_DEFINE for now.

File: subsys/bluetooth/controller/hal/cntr.h
Comment (cvinayak) [inline subsys/bluetooth/controller/hal/cntr.h:7]:
```suggestion
```

Include `#include <stdint.h>` in c files where needed. This is not right to have here to satisfy files that include `#include "hal/cntr.h"`

That said, I will have a future action item on myself to clean up other redundant/incorrect includes in controller hal folder, like sys/dlist.h etc  which is never used in cntr.c file.

**Review by cvinayak — DISMISSED** [review]

**Comment by andyross** [comment]
Resurrect this guy with a rebase, and a few extra patches that port the tree internals away from the old API, remove the old API entirely, and then remove the wrappers.  Should be good to go per existing comments.  We'll see what CI says.

File: subsys/bluetooth/controller/hal/cntr.h
Comment (andyross) [reply subsys/bluetooth/controller/hal/cntr.h:7]: Missed this comment.  There is a uint8_t and a uint32_t usage six lines down in this same file (the arguments to cntr_cmp_set()).  There is not way for this to compile correctly in all cases without this include here.

File: drivers/usb/device/usb_dc_kinetis.c
Comment (andyross) [reply drivers/usb/device/usb_dc_kinetis.c]: There's a port of that subsystem (and all the others) in this patch now.  But yes: in the old style you would specify how many objects you wanted to be able to allocate, and the overhead would be generated magically.  Now you specify how many bytes you want to use for heap storage and the overhead is included in that.

**Review by nashif — APPROVED** [review]

**Comment by andyross** [comment]
Heh, usb_dc_nrfx was one of the blind changes, isn't part of a default sanitycheck, but is part of CI.  Fixed typo, repushed.

**Comment by nashif** [comment]
nrf52_bsim is unrelated and caused by #29810

**Comment by andyross** [comment]
Next spin: now-unused variable warning in hid-cdc sample

**Pull request merged**
